File: jump-game-ii/jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int n = nums.length;
        if(n == 1) return 0;
        if(nums[0] == 0) return -1;
        int maxReach = nums[0];
        int minJump = 1;
        int lastStep = nums[0];
        int i=1;
        
        while(i < n-1){
            maxReach = Math.max(maxReach, nums[i] + i);
            lastStep--;
            
            if(lastStep == 0){
                minJump++;
                if(i >= maxReach) return -1;
                lastStep = maxReach - i;
            }
            i++;
        }
        
        return minJump;
    }
}

File: jump-game-ii/jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        
        if(n == 1) return 0;
        
        // if(nums[0] == 0) return -1;
        
        int lastStep = nums[0];
        
        int maxReachable = nums[0];
        
        int curr= 1;
        int minJumps = 1;        
        while(curr < n-1){
            maxReachable = max(maxReachable, nums[curr]+curr);
            if(lastStep == curr){
                if(maxReachable <= curr) return -1;
                lastStep = maxReachable;
                minJumps++;
            }
            curr++;
        }
        
        return minJumps;
    }
};